### PR TITLE
fix(prefill): validate custom_mask shape in single_prefill_with_kv_cache

### LINKED ISSUE #3176: Silent memory corruption when custom_mask shape doesn't match (qo_len, kv_len) in single_prefill_with_kv_cache
While vllm-omni was adding support for flashinfer a [data corruption/illegal access] issue was detected due to lack of validation of the `custom_mask`'s shape. The [functions documentation](https://docs.flashinfer.ai/generated/flashinfer.prefill.single_prefill_with_kv_cache.html) states
> custom_mask (Optional[torch.Tensor]) –
>
> The custom boolean mask tensor, shape: [qo_len, kv_len]. The elements in the mask tensor should be either True or False, where False means the corresponding element in the attention matrix will be masked out.

I am unable to provide a simple reproducer for illegal access. @lishunyang12 was able to produce it on end to end test of Hunyuan Video 1.5 with https://github.com/vllm-project/vllm-omni/pull/3079/changes/b25c755322cba84e14802e1f62acf5f982480618. A simpler reproducer that detects corruption is:
```
import torch
import flashinfer


def test_bad_mask_shape():
    qo_len, kv_len = 10, 20
    num_qo_heads, num_kv_heads = 8, 4
    head_dim = 128
    kwargs= {}

    q = torch.randn(qo_len, num_qo_heads, head_dim, dtype=torch.float16)
    k = torch.randn(kv_len, num_kv_heads, head_dim, dtype=torch.float16)
    v = torch.randn(kv_len, num_kv_heads, head_dim, dtype=torch.float16)

    bad_mask = torch.ones(qo_len, kv_len - 5, dtype=torch.bool)

    # Correct mask for comparison
    correct_mask = torch.ones(qo_len, kv_len, dtype=torch.bool)
    # Mask out the last 5 columns to make it meaningful
    correct_mask[:, 15:] = False

    print(f"Q shape: {q.shape} (qo_len={qo_len})")
    print(f"K shape: {k.shape} (kv_len={kv_len})")
    print(f"V shape: {v.shape}")
    print(f"Bad mask shape: {bad_mask.shape} (expected [{qo_len}, {kv_len}])")
    print(f"Correct mask shape: {correct_mask.shape}")
    print(f"Bad packed mask size: {(bad_mask.numel() + 7) // 8} bytes")
    print(f"Correct packed mask size: {(correct_mask.numel() + 7) // 8} bytes")
    print()

    print("Running with CORRECT mask...")
    kwargs["custom_mask"] = correct_mask
    o_correct = flashinfer.single_prefill_with_kv_cache(q, k, v, ** kwargs)
    print(f"Correct output (first head, first 5 tokens): {o_correct[0, 0, :5]}")

    # Run with bad mask
    print("\nRunning with BAD mask (shape mismatch)...")
    try:
        kwargs["custom_mask" = bad_mask
        o_bad = flashinfer.single_prefill_with_kv_cache(q, k, v, ** kwargs)
        print(f"Bad output (first head, first 5 tokens): {o_bad[0, 0, :5]}")
        
        # Check for corruption
        diff = (o_correct - o_bad).abs().max().item()
        print(f"\nMax absolute difference: {diff:.6f}")
        
        if diff > 1e-5:
            print("Results differ significantly!")
            
    except RuntimeError as e:
        print("This is the kernel detecting the illegal access.")


if __name__ == "__main__":
    test_bad_mask_shape()
```

### PR DESCRIPTION
**Pull request by IamYipi**
## Problem

`single_prefill_with_kv_cache` packs `custom_mask` with

```python
packed_custom_mask = packbits(custom_mask.contiguous().view(-1), bitorder="little")
```

`view(-1)` flattens any shape without checking its size, so a mask that is not
`(qo_len, kv_len)` is accepted. When the mask is smaller than the attention
matrix the kernel reads past the end of the packed buffer and the call returns
corrupted results **with no error at all**.

Fixes #3176.

## Evidence

Reproduced on an RTX 6000 Ada (sm_89), `qo_len=10, kv_len=20, num_qo_heads=8,
num_kv_heads=4, head_dim=128`.

A `(10, 15)` mask needs `ceil(10*20/8) = 25` packed bytes but only supplies
`ceil(10*15/8) = 19`, so **50 bits are read out of bounds**.

The result of the malformed call matches neither of the two legitimate outputs,
which confirms the values come from unrelated memory rather than from some
benign fallback:

| comparison | max abs diff |
|---|---|
| malformed mask vs no mask | 1.301 |
| malformed mask vs all-`True` mask | 1.301 |
| malformed mask vs correct mask | 1.417 |
| correct mask vs all-`True` mask (control: the mask does change the output) | 1.225 |

The same silent acceptance happens for a mask with too many columns, with too
few rows, and for a transposed mask.

## Fix

Validate the shape before packing. `qo_len` comes from `q.size(0)`, and `kv_len`
from `k.size(0)` or `k.size(1)` depending on `kv_layout`.

The flattened spelling packs to exactly the same bytes, so a 1-D mask with
`qo_len * kv_len` elements is still accepted and existing callers that rely on
it keep working.

## Tests

Added to `tests/attention/test_single_prefill.py`:

- `test_single_prefill_custom_mask_shape_mismatch` — 4 malformed shapes
  (fewer columns, more columns, fewer rows, transposed) x 2 `kv_layout`s,
  each expected to raise `ValueError`. **All 8 fail without the fix.**
- `test_single_prefill_custom_mask_valid_shapes_accepted` — 2-D and flattened
  masks x 2 `kv_layout`s, guarding against over-strict validation.

```
12 passed, 218 deselected           # -k custom_mask, after the fix
8 failed, 4 passed                  # same selection, before the fix
4 passed                            # regression subset on the unmasked path
```

`ruff check` and `ruff format` are clean on both files.

## Note, out of scope for this PR

`flashinfer/pod.py:603` and `flashinfer/pod.py:1227` pack `custom_mask` the same
way and look like they have the same gap. I left them untouched to keep this
change small; happy to follow up if you would like them covered too.

## Summary by CodeRabbit

* **Bug Fixes**
  * Added validation for custom attention mask dimensions before processing.
  * Invalid masks, including incorrectly ranked or shaped masks, now produce a clear error instead of being processed incorrectly.
  * Correctly sized masks continue to work in both two-dimensional and flattened one-dimensional formats.

* **Tests**
  * Added coverage for valid and invalid mask shapes across supported attention layouts.

### CODE REVIEW

File: flashinfer/prefill.py [hunk 1]
```diff
@@ -1455,6 +1455,22 @@ def single_prefill_with_kv_cache(
...
+        qo_len = q.size(0)
+        kv_len = k.size(0) if kv_layout == "NHD" else k.size(1)
+        # packbits() below flattens the mask without looking at its size, so a mask
+        # smaller than the attention matrix makes the kernel read past the end of the
+        # packed buffer and silently return corrupted results. The flattened spelling
+        # packs to the same bytes, so it stays accepted.
+        if custom_mask.dim() == 2:
+            mask_ok = tuple(custom_mask.shape) == (qo_len, kv_len)
+        else:
+            mask_ok = custom_mask.numel() == qo_len * kv_len
```
[the remark this answers is not in the document]
Comment (IamYipi) [reply flashinfer/prefill.py]: Confirmed and fixed in de9b0d0. A rank-3 mask such as `(1, qo_len, kv_len)` was reaching the element-count branch and being accepted. It packs to the same bytes so the result was still correct, but it let a caller hand a batched-looking mask to an API with no batch dimension and get no diagnostic — the exact failure mode this validation is meant to remove.

The rank check is now explicit: 1-D with `qo_len * kv_len` elements, or 2-D matching `(qo_len, kv_len)`. Every other rank is rejected. Added `(1, 10, 20)` to the mismatch parametrization; the selection is now 14 passed.

**Comment by IamYipi** [comment]
Updated after review. Three commits now:

1. `cc9d577` — the shape validation itself.
2. `de9b0d0` — addresses the CodeRabbit finding: the first version routed every non-2-D mask through the element-count branch, so a rank-3 `(1, qo_len, kv_len)` mask was still accepted. The rank check is now explicit — 1-D with `qo_len * kv_len` elements, or 2-D matching `(qo_len, kv_len)`; every other rank is rejected, with `(1, 10, 20)` added to the parametrization.
3. `e7262ba` — documents the two accepted spellings in the `custom_mask` docstring.

### Local verification

Run on an RTX 6000 Ada (sm_89), CUDA 12.8 container, torch 2.14.0+cu130, editable install of this branch:

| suite | result |
|---|---|
| `tests/attention/test_single_prefill.py` (full) | 220 passed, 12 skipped |
| `-k custom_mask` | 14 passed (8 of them fail without the fix) |
| `tests/attention/test_block_sparse.py` + `test_sliding_window.py` | 13832 passed, 8976 skipped |
| `ruff check` / `ruff format --check` | clean |

I ran block-sparse and sliding-window specifically because they are the callers that build a dense `custom_mask` and pass it to this function, so they were the realistic regression risk for a new shape check. `cascade.py` calls `single_prefill_with_kv_cache` without a custom mask, so it is unaffected.

### CI

`Test Results Summary` is red because the run stops at `CI skipped (pending authorization)` — external PRs need a maintainer or `ci-users` member to comment `@flashinfer-bot run`. Whenever someone has a moment to authorize it, I will fix anything it surfaces. Every other check is green.
